components/notifications-service/integration_test/notification_test.go

```diff
@@ -0,0 +1,1207 @@
+		assert.Equal(t, "plan9", supportsActual.OSName)
+		assert.Equal(t, "Esoterica", supportsActual.OSFamily)
+		assert.Equal(t, "99", supportsActual.InSpec)
+
+		require.Len(t, profileActual.Attributes, 1)
+		attrActual := profileActual.Attributes[0]
+
+		assert.Equal(t, "example_attr", attrActual.Name)
+		assert.Equal(t, "example attribute in an integration test", attrActual.Options.Description)
+
+		require.Len(t, profileActual.Controls, 1)
+		controlActual := profileActual.Controls[0]
+
+		assert.Equal(t, "Example Control", controlActual.ID)
+		assert.Equal(t, "Simulated Failure", controlActual.Title)
+		assert.Equal(t, "An example control from an integration_test", controlActual.Desc)
+		assert.Equal(t, "failed", controlActual.Status)
+		assert.Equal(t, "halt -p", controlActual.Code)
+		assert.Equal(t, 2, controlActual.NumberOfTests)
+		assert.Equal(t, 1, controlActual.NumberOfFailedTests)
+		assert.Equal(t, 7.5, controlActual.Impact)
+
+		assert.Equal(t, "example_profile.rb", controlActual.SourceLocation.Ref)
+		assert.Equal(t, 5, controlActual.SourceLocation.Line)
+
+		require.Len(t, controlActual.Refs, 1)
+		refsActual := controlActual.Refs[0]
+		assert.Equal(t, "http://compliancerules.example.net", refsActual.URL)
+		assert.Equal(t, "http://compliancerules.example", refsActual.URI)
+
+		require.Len(t, controlActual.Results, 1)
+		resultActual := controlActual.Results[0]
+
+		assert.Equal(t, "failed", resultActual.Status)
+		assert.Equal(t, "EXAMPLE CONTROL FAILED MESSAGE", resultActual.Message)
+		assert.Equal(t, "example skip message", resultActual.SkipMessage)
+		assert.Equal(t, "not good", resultActual.CodeDesc)
+		assert.Equal(t, 3.5, resultActual.RunTime)
+
+		resultStartTime, err := time.Parse(time.RFC3339Nano, resultActual.StartTime)
+		assert.NoError(t, err)
+		assert.Equal(t, startTime, resultStartTime)
+	})
+}
+
+func TestNotificationDispatchWithMultipleDestinations(t *testing.T) {
+	err := suite.DeleteEverything()
+	require.NoError(t, err)
+
+	ts := newTestServer()
+	defer ts.Close()
+
+	addRules(t,
+		slackAlertRuleFor(api.Rule_CCRFailure, ts.SlackURL()),
+		slackAlertRuleFor(api.Rule_ComplianceFailure, ts.SlackURL()),
+		serviceNowAlertRuleFor(api.Rule_CCRFailure, ts.ServiceNowURL(), ""),
+		serviceNowAlertRuleFor(api.Rule_ComplianceFailure, ts.ServiceNowURL(), ""),
+		webhookAlertRuleFor(api.Rule_CCRFailure, ts.WebhookURL()),
+		webhookAlertRuleFor(api.Rule_ComplianceFailure, ts.WebhookURL()),
+	)
+	t.Run("a CCR failure sends one alert to each destination", func(t *testing.T) {
+		_, err := suite.Client.Notify(ctx, ccrFailureEvent())
+		require.NoError(t, err)
+		actualURLPaths := make(map[string]bool)
+		for i := 0; i < 3; i++ {
+			postData := ts.GetLastPost()
+			assert.NotNil(t, postData)
+			actualURLPaths[postData.URL] = true
+		}
+		expectedURLPaths := map[string]bool{
+			"/slackalert/":      true,
+			"/servicenowalert/": true,
+			"/webhookalert/":    true,
+		}
+
+		assert.Equal(t, expectedURLPaths, actualURLPaths)
```

Comment:
Is order of the alerts really significant? (But then again, is it really worth writing a test to check for order independence? 🤷 )

Reply:
This test should not depend on the order of the alerts, that's why I accumulated the URLs into a map. The existing system may accidentally produce a stable order (I haven't checked at all) but perhaps the stable order is brittle (e.g., order of rows returned from postgres when no sort is specified is often "stable" until you make an unrelated schema change)